spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
+    // it assumes signature verification so it can be computationally expensive
+    func votingPowerIn(v1 []Validator, v2 ValidatorSet) int64
 
-Formally,
-\[
-\sum*{(v,p) \in h.Header.NextV \wedge correct(v,h.Header.bfttime + tp)} p >
-2/3 \sum*{(v,p) \in h.Header.NextV} p
-\]
+    // add this state as trusted to the store
+    func add(store Store, trustedState TrustedState) error
 
-_Assumption_: "correct" is defined w.r.t. realtime (some Newtonian global notion of time, i.e., wall time), while _bfttime_ corresponds to the reading of the local clock of a validator (how this time is computed may change when the Tendermint consensus is modified). In this note, we assume that all clocks are synchronized to realtime. We can make this more precise eventually (incorporating clock drift, accuracy, precision, etc.). Right now, we consider this assumption sufficient, as clock synchronization (under NTP) is in the order of milliseconds and _tp_ is in the order of weeks.
+    // retrieve the trusted state at given height if it exists (error = nil)
+    // return an error if there are no trusted state for the given height
+    // if height = 0, return the latest trusted state
+    func get(store Store, height int64) (TrustedState, error)
+```
 
-_Remark_: This failure model might change to a hybrid version that takes heights into account in the future.
+### Failure Model
 
-The specification in this document considers an implementation of the lite client under this assumption. Issues like _counter-factual signing_ and _fork accountability_ and _evidence submission_ are mechanisms that justify this assumption by incentivizing validators to follow the protocol.
-If they don't, and we have more that 1/3 faults, safety may be violated. Our approach then is to _detect_ these cases (after the fact), and take suitable repair actions (automatic and social). This is discussed in an upcoming document on "Fork accountability". (These safety violations include the lite client wrongly trusting a header, a fork in the blockchain, etc.)
+The lite client specification is defined with respect to the following failure model: If a block `b` is generated
+at time `Time` (and this time is stored in the block), then a set of validators that hold more than 2/3 of the voting
+power in `validators(b.Header.NextValidatorsHash)` is correct until time `b.Header.Time + TRUSTED_PERIOD`.
 
-## Lite Client Trusting Spec
+*Assumption*: "correct" is defined w.r.t. realtime (some Newtonian global notion of time, i.e., wall time),
+while `Header.Time` corresponds to the [BFT time](bft-time.md). In this note, we assume that clocks of correct processes
+are synchronized (for example using NTP), and therefore there is bounded clock drift (CLOCK_DRIFT) between local clocks and
+BFT time. More precisely, for every correct process p and every header (correctly generated by the Tendermint consensus)
```

Comment:
This sentence is a bit hard to follow with the two brackets, maybe we can combine them like `and every header.Time (ie. BFT Time, for a header correctly generated by the Tendermint consensus), the following ...`